### PR TITLE
Add ability to visualize the state lattice tree expansions

### PR DESCRIPTION
**Pull request by embeddedadam**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | (None) |
| Primary OS tested on | (Ubuntu) |
| Robotic platform tested on | (None) |
| Does this PR contain AI generated software? | (No) |

---

## Description of contribution in a few bullet points
Added debug flag in smac_planner_lattice.cpp that enables visualization of state lattice tree expansions

## Description of documentation updates required from your changes
* Added new parameter in smac_planner_lattice _debug_visualizations
* Added visualisation like in hybrid A* planner after _debug_visualizations is set

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_smac_planner/src/smac_planner_lattice.cpp [hunk 1]
```diff
@@ -282,11 +298,27 @@ nav_msgs::msg::Path SmacPlannerLattice::createPlan(
...
+  std::unique_ptr<std::vector<std::tuple<float, float, float>>> expansions = nullptr;
+  if (_debug_visualizations) {
+    expansions = std::make_unique<std::vector<std::tuple<float, float, float>>>();
+  }
   // Note: All exceptions thrown are handled by the planner server and returned to the action
   if (!_a_star->createPath(
       path, num_iterations, _tolerance / static_cast<float>(_costmap->getResolution())))
   {
+    if (_debug_visualizations) {
+      geometry_msgs::msg::PoseArray msg;
```
Comment (SteveMacenski) [inline nav2_smac_planner/src/smac_planner_lattice.cpp:310]: These 2 blocks are not covered by any unit or system tests, we need to have that in order to merge the PR for quality control requirements

**Review by SteveMacenski — COMMENTED** [review]
Can you show what this output looks like in rviz?

File: nav2_smac_planner/src/smac_planner_lattice.cpp (hunk 1 above)
Comment (SteveMacenski) [reply nav2_smac_planner/src/smac_planner_lattice.cpp:310]: The WPF test failed oddly, so I retriggered CI to see if that was just a fluke

**Comment by SteveMacenski** [comment]
The new parameter needs to go into the readme / [configuration guide](https://navigation.ros.org/configuration/packages/smac/configuring-smac-lattice.html)

**Comment by mikeferguson** [comment]
As a general note, that lattice might become VERY large ... and very large MarkerArrays can definitely tank performance in RVIZ.

In most cases where I've seen visualization of a state lattice tree, it has been done with a OccupancyGrid type message, which simply shows which cells have had related states expanded within them. This definitely works for debugging things like "my heuristic is messed up and I'm expanding states in the wrong place" which I feel is one of the most common issues, but wouldn't help as much in debugging heading-specific issues.

**Review by mikeferguson — COMMENTED** [review]
As a general note, that lattice might become VERY large ... and very large MarkerArrays can definitely tank performance in RVIZ.

In most cases where I've seen visualization of a state lattice tree, it has been done with a OccupancyGrid type message, which simply shows which cells have had related states expanded within them. This definitely works for debugging things like "my heuristic is messed up and I'm expanding states in the wrong place" which I feel is one of the most common issues, but wouldn't help as much in debugging heading-specific issues.

**Comment by SteveMacenski** [comment]
Huh, that's an interesting idea. 

I can say that the Hybrid-A* expansions show up fine but starts to lag down rviz a _little bit._ This would introduce comparatively less of those since the individual primitives are so much longer in the state lattices than in Hybrid-A*. Right now as the PR stands, it only does the end-points of the lattices, not the intermediary poses that are used to populate the normalized dense path. I think due to that, and including the orientation information which can be important debugging information, I think this PR is OK as it is using the markers.

I definitely don't disagree though that if we want to publish the full lattice samples that we should use the method you describe. Basically that would just entail iterating over them and marking each cell a primitive lands in, which should be straight forward to implement. I think that would be a lovely additional capability which shows another part of the expansion pattern which might yield insights.

**Comment by SteveMacenski** [comment]
@embeddedadam this PR now has commits from the main branch - this is not rebased / merged in correctly to only have the software you've created for this PR in your tree. I'm not able to review this until that is resolved. Try rebasing or merging in changes more standardly. 

Also, please provide the rviz screen shot I requested for what this looks like! :smile:

**Comment by SteveMacenski** [comment]
Hey! Any update? I'd love to have this in :-)

**Comment by SteveMacenski** [comment]
Working on this now to bring it home - PR to open shortly

**Pull request closed without merging**